<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Hugo 0.84.4" />


<title>Visualize the &#39;real&#39; circular genome - Zuguang Gu</title>
<meta property="og:title" content="Visualize the &#39;real&#39; circular genome - Zuguang Gu">



  








<link href='//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css' rel='stylesheet' type='text/css' />



<link rel="stylesheet" href="/css/fonts.css" media="all">
<link rel="stylesheet" href="/css/main.css" media="all">



  </head>
  <body>
    <div class="wrapper">
      <header class="header">
        <nav class="nav">
  <a href="/" class="nav-logo-text">
A Bioinformagician
  </a>

  <ul class="nav-links">
    
    <li><a href="/about/">About</a></li>
    
    <li><a href="/post/">Post</a></li>
    
    <li><a href="/software/">Software</a></li>
    
    <li><a href="/publication/">Publication</a></li>
    
    <li><a href="/talks/">Talks</a></li>
    
    <li><a href="https://github.com/jokergoo">GitHub</a></li>
    
    <li><a href="/cv/">CV</a></li>
    
  </ul>
</nav>

      </header>


<main class="content" role="main">

  <article class="article">
    
    <span class="article-duration">3 min read</span>
    

    <h1 class="article-title">Visualize the &#39;real&#39; circular genome</h1>

    
    <span class="article-date">2021-09-18</span>
    

    <div class="article-content">
      
<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<p><strong>circlize</strong> applies circular visualization. Essentially, it visualizes data in categories. E.g. for genomic data,
the circle is segmented and each segment (or called sector) corresponds to one single chromosome. The following code visualizes 24 chromosomes of human genome.</p>
<pre class="r"><code>library(circlize)
circos.initializeWithIdeogram()</code></pre>
<p><img src="/post/2021-09-18-visualize-the-real-circular-genome_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Now the question is how to visualize a “real” circular genome where the complete circle corresponds to the genome and the “end” of the genome
goes smoothly over the “start” of the genome.</p>
<p>Maybe we can create a single sector and set <code>gap.degree</code> to 0. In the following, I tried to draw a line from <code>x = 9</code> to <code>x = 1</code>. In
the circular genome, we expect the line goes through <code>x = 0</code>.</p>
<pre class="r"><code>circos.par(gap.degree = 0, cell.padding = c(0.02, 0, 0.02, 0))
circos.initialize(&quot;foo&quot;, xlim = c(0, 10))
circos.track(ylim = c(0, 1))
circos.lines(c(9, 1), c(0.2, 0.2), col = &quot;blue&quot;)
circos.arrow(9, 1, y = 0.7, width = 0.2)</code></pre>
<p><img src="/post/2021-09-18-visualize-the-real-circular-genome_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<p>Unfortunately it is not what we expected. If a line is drawn from <code>x = 9</code> to <code>x = 1</code> in the normal mode, it only goes reverse close wisely
on the circle. And there is also a radical line in the ring which is actually the border of <code>x = 0</code> and <code>x = 10</code>. Thus in the coordination
system, <code>x = 1</code> and <code>x = 9</code> are still far from each other.</p>
<p>To support the “real” visualization on circular genomes, in <strong>circlize</strong>, now I add a new mode called “the ring mode”.
Simply add <code>ring = TRUE</code> in <code>circos.initialize()</code>.</p>
<pre class="r"><code>circos.initialize(&quot;foo&quot;, xlim = c(0, 10), ring = TRUE)
circos.track(ylim = c(0, 1))
circos.lines(c(9, 1), c(0.2, 0.2), col = &quot;blue&quot;)
circos.arrow(9, 1, y = 0.7, width = 0.2)</code></pre>
<p><img src="/post/2021-09-18-visualize-the-real-circular-genome_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<p>In the next example, I visualize genome of
the Human gammaherpesvirus 4. The genomic coordinates of genes are from <a href="https://www.ncbi.nlm.nih.gov/genome/browse/#!/proteins/10261/459250%7CHuman%20gammaherpesvirus%204/" class="uri">https://www.ncbi.nlm.nih.gov/genome/browse/#!/proteins/10261/459250%7CHuman%20gammaherpesvirus%204/</a>.
As you can see here, start position in the first row in larger than the end position.</p>
<pre class="r"><code>df = read.csv(&quot;~/Downloads/proteins_10261_459250.csv&quot;)
head(df)</code></pre>
<pre><code>##          X.Name   Accession  Start  Stop Strand  GeneID   Locus    Locus.tag
## 1 viral segment NC_007605.1 166103  1680      + 3783751  LMP-2A  HHV4_LMP-2A
## 2 viral segment NC_007605.1     59  1680      + 3783760  LMP-2B  HHV4_LMP-2B
## 3 viral segment NC_007605.1   1736  5692      + 3783722   BNRF1   HHV4_BNRF1
## 4 viral segment NC_007605.1   9675 10187      + 3783689   BCRF1 HHV4_BCRF1.1
## 5 viral segment NC_007605.1  12541 13692      + 3783691   BWRF1 HHV4_BWRF1.1
## 6 viral segment NC_007605.1  14409 35694      + 3783746 EBNA-LP HHV4_EBNA-LP
##   Protein.product Length            Protein.Name
## 1     YP_401631.1    497 membrane protein LMP-2A
## 2     YP_401632.1    378 membrane protein LMP-2B
## 3     YP_401633.1   1318    tegument protein G75
## 4     YP_401634.1    170    interleukin-10 BCRF1
## 5     YP_401635.1    383  protein BWRF1, partial
## 6     YP_401636.1    506 nuclear antigen EBNA-LP</code></pre>
<pre class="r"><code>df = df[, 2:5]</code></pre>
<p>I wrote a helper function <code>circos.initializeCircularGenome()</code> especially for circular genomes.
In other parts of the code, just use <strong>circlize</strong> in the normal way.</p>
<pre class="r"><code>circos.initializeCircularGenome(&quot;NC_007605.1&quot;, genome_size = 171823)
circos.genomicTrack(df, ylim = c(0, 1), panel.fun = function(region, value, ...) {
    circos.genomicLines(region, value = runif(nrow(region)), type = &quot;segment&quot;, 
        col = ifelse(value[, 1] == &quot;+&quot;, &quot;Red&quot;, &quot;blue&quot;))
})
circos.genomicLink(df[1, ], df[60, ], col = &quot;#00FF0080&quot;)</code></pre>
<p><img src="/post/2021-09-18-visualize-the-real-circular-genome_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;" /></p>

    </div>
  </article>

  
  <script src="https://utteranc.es/client.js"
    repo="jokergoo/jokergoo.github.io"
    issue-term="pathname"
    theme="github-light"
    crossorigin="anonymous"
    async>
</script>
  
  
</main>

      <footer class="footer">
        <ul class="footer-links">
          <li>
            <a href="/index.xml" type="application/rss+xml" target="_blank">RSS feed</a>
          </li>
          <li>
            <a href="https://gohugo.io/" class="footer-links-kudos">Made with <img src="/images/hugo-logo.png" alt="Img link to Hugo website" width="22" height="22"></a>
          </li>
        </ul>
      </footer>

    </div>
    



<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>



<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.configure({languages: []}); hljs.initHighlightingOnLoad();</script>



    

    
  </body>
</html>

