---
title: "Reverse x-axes in the circular plot"
date: 2020-08-17
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<p>By default, in <em>circlize</em>, the directions of x-axes of all cells are clockwise, which means, in the cells
located on top of the circle, the direction of x-axes is from left to right, which fits how we normally
read the figures. For example, the following figure shows several tracks with different graphics.</p>
<pre class="r"><code>library(circlize)
circos.par(gap.degree = 4)
circos.initialize(letters[1:8], xlim = c(0, 10))
circos.track(ylim = c(0, 10), panel.fun = function(x, y) {
    circos.axis()
    circos.points(0:10, 0:10)
    circos.text(CELL_META$xlim[1] + mm_x(5), CELL_META$ycenter, CELL_META$sector.index, col = &quot;red&quot;)
})
circos.track(ylim = c(0, 10), panel.fun = function(x, y) {
    circos.rect(1, 4, 5, 6)
    circos.arrow(6, 9)
})
circos.link(&quot;a&quot;, 1, &quot;c&quot;, 9)</code></pre>
<p><img src="/post/2020-08-17-reverse-x-axes-in-the-circular-plot_files/figure-html/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<p>However, some users might want the axes in the other direction, like the request <a href="https://github.com/jokergoo/circlize/issues/124">here</a>. The solution was to manually reverse the x-axes by defining a <code>rev_x()</code> function.</p>
<p>Then all positions related to x-axes should be wrapped by <code>rev_x()</code>.</p>
<pre class="r"><code>rev_x = function(x, sector.index = NULL) {
    if(!is.null(sector.index)) set.current.cell(sector.index, 1)

    xrange = CELL_META$xlim
    xrange[2] - x + xrange[1]
}
circos.par(gap.degree = 4)
circos.initialize(letters[1:8], xlim = c(0, 10))
circos.track(ylim = c(0, 10), panel.fun = function(x, y) {
    circos.axis(major.at = rev_x(seq(0, 10, by = 2)), labels = seq(0, 10, by = 2))
    circos.points(rev_x(0:10), 0:10)
    circos.text(rev_x(CELL_META$xlim[1] + mm_x(5)), CELL_META$ycenter, CELL_META$sector.index, col = &quot;red&quot;)
})
circos.track(ylim = c(0, 10), panel.fun = function(x, y) {
    circos.rect(rev_x(1), 4, rev_x(5), 6)
    circos.arrow(rev_x(6), rev_x(9))
})
circos.link(&quot;a&quot;, rev_x(1, &quot;a&quot;), &quot;c&quot;, rev_x(9, &quot;c&quot;))</code></pre>
<p><img src="/post/2020-08-17-reverse-x-axes-in-the-circular-plot_files/figure-html/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<p>This might be painful if there are so many positions on x-axes to adjust. From circlize 0.4.11. I have added
a new global option <code>xaxis.clock.wise</code> to globally control the directions of x-axes. The <code>xaxis.clock.wise</code>
option should be called by <code>circos.par()</code> and should be called before the circular plot is initialized.</p>
<p>See the following example. It is super easy to reverse all x-axes now.</p>
<pre class="r"><code>circos.par(gap.degree = 4, xaxis.clock.wise = FALSE)
circos.initialize(letters[1:8], xlim = c(0, 10))
circos.track(ylim = c(0, 10), panel.fun = function(x, y) {
    circos.axis()
    circos.points(0:10, 0:10)
    circos.text(CELL_META$xlim[1] + mm_x(5), CELL_META$ycenter, CELL_META$sector.index, col = &quot;red&quot;)
})
circos.track(ylim = c(0, 10), panel.fun = function(x, y) {
    circos.rect(1, 4, 5, 6)
    circos.arrow(6, 9)
})
circos.link(&quot;a&quot;, 1, &quot;c&quot;, 9)</code></pre>
<p><img src="/post/2020-08-17-reverse-x-axes-in-the-circular-plot_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<p>Setting <code>circos.par(xaxis.clock.wise = FALSE)</code> reverses all positions on x-direction. Following is a genomic
circular plot with and without reversing x-axes. As you can see, the code for making the plot are exactly the same.
The only difference is in the second plot, we change the direction of x-axes by setting <code>circos.par(xaxis.clock.wise = FALSE)</code>.</p>
<pre class="r"><code>load(system.file(package = &quot;circlize&quot;, &quot;extdata&quot;, &quot;DMR.RData&quot;))
bed_list = list(DMR_hyper, DMR_hypo)

par(mfrow = c(1, 2))
circos.initializeWithIdeogram()
circos.genomicRainfall(bed_list, pch = 16, cex = 0.4, col = c(&quot;#FF000080&quot;, &quot;#0000FF80&quot;))
circos.genomicDensity(DMR_hyper, col = c(&quot;#FF000080&quot;), track.height = 0.1)
circos.genomicDensity(DMR_hypo, col = c(&quot;#0000FF80&quot;), track.height = 0.1)
circos.clear()

circos.par(xaxis.clock.wise = FALSE)
circos.initializeWithIdeogram()
circos.genomicRainfall(bed_list, pch = 16, cex = 0.4, col = c(&quot;#FF000080&quot;, &quot;#0000FF80&quot;))
circos.genomicDensity(DMR_hyper, col = c(&quot;#FF000080&quot;), track.height = 0.1)
circos.genomicDensity(DMR_hypo, col = c(&quot;#0000FF80&quot;), track.height = 0.1)</code></pre>
<p><img src="/post/2020-08-17-reverse-x-axes-in-the-circular-plot_files/figure-html/unnamed-chunk-5-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<p>We show a second genomic circular plot which visualizes all transcripts of three genes. We first define
a <code>draw_tx()</code> function which draws the plot and we will use it twice.</p>
<pre class="r"><code>draw_tx = function() {
    tp_family = readRDS(system.file(package = &quot;circlize&quot;, &quot;extdata&quot;, &quot;tp_family_df.rds&quot;))
    circos.genomicInitialize(tp_family)
    circos.track(ylim = c(0, 1), 
        bg.col = c(&quot;#FF000040&quot;, &quot;#00FF0040&quot;, &quot;#0000FF40&quot;), 
        bg.border = NA, track.height = 0.05)
    n = max(tapply(tp_family$transcript, tp_family$gene, function(x) length(unique(x))))
    circos.genomicTrack(tp_family, ylim = c(0.5, n + 0.5), 
        panel.fun = function(region, value, ...) {
            all_tx = unique(value$transcript)
            for(i in seq_along(all_tx)) {
                l = value$transcript == all_tx[i]
                # for each transcript
                current_tx_start = min(region[l, 1])
                current_tx_end = max(region[l, 2])
                circos.lines(c(current_tx_start, current_tx_end), 
                    c(n - i + 1, n - i + 1), col = &quot;#CCCCCC&quot;)
                circos.genomicRect(region[l, , drop = FALSE], ytop = n - i + 1 + 0.4, 
                    ybottom = n - i + 1 - 0.4, col = &quot;orange&quot;, border = NA)
            }
    }, bg.border = NA, track.height = 0.4)
}</code></pre>
<p>And we make the plot:</p>
<pre class="r"><code>par(mfrow = c(1, 2))
draw_tx()
circos.clear()

circos.par(xaxis.clock.wise = FALSE)
draw_tx()</code></pre>
<p><img src="/post/2020-08-17-reverse-x-axes-in-the-circular-plot_files/figure-html/unnamed-chunk-7-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<p>If the reversing of x-axes is not applied globally, but rather a subset of cells, <code>rev_x()</code> should
be manually applied to those cells where x-axes are reversed. In the following example, we only reverse
the x-axis in the top cell.</p>
<pre class="r"><code>circos.par(gap.degree = 5)
circos.initialize(letters[1:2], xlim = c(0, 10))
circos.track(ylim = c(0, 10))
circos.arrow(1, 9, sector.index = &quot;a&quot;)
circos.axis(major.at = seq(0, 10, by = 2), sector.index = &quot;a&quot;)

circos.arrow(rev_x(1), rev_x(9), sector.index = &quot;b&quot;)
circos.axis(major.at = rev_x(seq(0, 10, by = 2)), labels = seq(0, 10, by = 2), sector.index = &quot;b&quot;)</code></pre>
<p><img src="/post/2020-08-17-reverse-x-axes-in-the-circular-plot_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>circos.clear()</code></pre>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
<script>
  $(document).ready(function(){
  $("main.content").css("max-width", "900px");
});
</script>
